lib/unix.bzl

```diff
@@ -0,0 +1,326 @@
+    "strings",
+    "strip",
+    "stty",
+    "tabs",
+    "tail",
+    "talk",
+    "tee",
+    "test",
+    "time",
+    "touch",
+    "tput",
+    "tr",
+    "true",
+    "tsort",
+    "tty",
+    "type",
+    "ulimit",
+    "umask",
+    "unalias",
+    "uname",
+    "uncompress",
+    "unexpand",
+    "unget",
+    "uniq",
+    "unlink",
+    "uucp",
+    "uudecode",
+    "uuencode",
+    "uustat",
+    "uux",
+    "val",
+    "vi",
+    "wait",
+    "wc",
+    "what",
+    "who",
+    "write",
+    "xargs",
+    "yacc",
+    "zcat",
+]
+
+TOOLCHAIN_TYPE = "@bazel_skylib//toolchains/unix:toolchain_type"
+MAKE_VARIABLES = "@bazel_skylib//toolchains/unix:make_variables"
+
+def _unix_toolchain_impl(ctx):
+    commands = {}
+    for cmd in _commands:
+        cmd_path = getattr(ctx.attr, cmd, None)
+        if not cmd_path:
+            cmd_path = None
+        commands[cmd] = cmd_path
+    cmd_paths = {
+        paths.dirname(cmd_path): None
+        for cmd_path in commands.values()
+        if cmd_path
+    }.keys()
+    return [platform_common.ToolchainInfo(
+        commands = commands,
+        paths = cmd_paths,
+    )]
+
+unix_toolchain = rule(
+    attrs = {
+        cmd: attr.string(
+            doc = "Absolute path to the {} command.".format(cmd),
+            mandatory = False,
+        )
+        for cmd in _commands
+    },
+    doc = """
+A toolchain capturing standard Unix shell commands.
+
+Provides:
+  commands: Dict of String, an item per command holding the path to the executable or None.
+  paths: List of String, deduplicated bindir paths. Suitable for generating `$PATH`.
+
+Use `unix_configure` to create an instance of this toolchain.
+See `unix_make_variables` on how to use this toolchain in genrules.
+""",
+    implementation = _unix_toolchain_impl,
+)
+
+def _unix_make_variables_impl(ctx):
+    toolchain = ctx.toolchains[TOOLCHAIN_TYPE]
+    cmd_vars = {
+        "UNIX_%s" % cmd.upper(): cmd_path
+        for cmd in _commands
+        for cmd_path in [toolchain.commands[cmd]]
+        if cmd_path
+    }
+    return [platform_common.TemplateVariableInfo(cmd_vars)]
+
+unix_make_variables = rule(
+    doc = """
+Provides Unix toolchain commands as custom make variables.
+
+Make variables:
+  Provides a make variable of the form `UNIX_<COMMAND>` for each available
+  command, where `<COMMAND>` is the name of the command in upper case.
+
+Use `unix.MAKE_VARIABLES` instead of instantiating this rule yourself.
+
+Example:
+  >>> genrule(
+          name = "use-grep",
+          srcs = [":some-file"],
+          outs = ["grep-out"],
+          cmd = "$(UNIX_GREP) search $(execpath :some-file) > $(OUTS)",
+          toolchains = [unix.MAKE_VARIABLES],
+      )
+""",
+    implementation = _unix_make_variables_impl,
+    toolchains = [TOOLCHAIN_TYPE],
+)
+
+def _unix_config_impl(repository_ctx):
+    cpu = get_cpu_value(repository_ctx)
+    commands = {}
+    for cmd in _commands:
+        cmd_path = repository_ctx.which(cmd)
```

Comment:
If we have a special toolchain configure rule, should we allow any form of user overrides, e.g., an environment variable with additional directories to look for, before looking into `PATH`, environment variables specifying the path to a tool, if I don't want to use the first on path, a mapping of alternative names.

The use case I have in mind is that authors often (accidentally) use non-POSIX extensions of their tools, so I often might have to use a particular version of the tool, which on my system might be called `gfind`, `gtest`, etc.

Reply:
That's a good point. I've implemented this as follows: Users can override individual commands by specifying `UNIX_<COMMAND>` environment variables. E.g. `UNIX_MAKE=/usr/bin/gmake`.

> e.g., an environment variable with additional directories to look for

Unfortunately, that's not so easy as `repository_ctx.which` doesn't allow to configure the search path, and there's no `repository_ctx.is_executable_file(path)`.